### PR TITLE
[Mips] Fix mfhi/mflo hazard miscompilation about div and mult

### LINKED ISSUE #81291: [mips] mfhi/mflo hazard miscompilation
```
int div(int a, int b, int c) {
  return a / b / c;
}
```
compiled with clang --target=mipsel-freestanding -mcpu=mips3 -O3 -fomit-frame-pointer
```
	div	$zero, $4, $5
	teq	$5, $zero, 7
	teq	$6, $zero, 7
	mflo	$1
	div	$zero, $1, $6
	jr	$ra
	mflo	$2
```

The earlier mips ISAs require a minimum of 2 instructions between a mflo/mfhi and the next mul/div instruction.  The output from clang V18.1.0 does not honor this.  It is also more than happy to put the mflo into the delay slot for the return, which can also cause bugs.

The reason for this hazard is the div/mul instructions start an asynchronous operation.  If an interrupt occurs the mfhi/mflo instruction might be flushed after the next div/mul has been started.  

GCC correctly handles this in both cases.

I'm more than happy to work on this if anyone is willing to give me a few pointers.  Forbidding mfhi/mflo in branch delay slots seems like a reasonable start, but I'm not sure how to handle the other case. I could add a pass to insert nops similar to the mulmul fix.  I'm also not completely certain when the constraint goes away, but it's at least present in mips1-4.

### PR DESCRIPTION
**Pull request by yingopq**
Fix issue1: In mips1-4, require a minimum of 2 instructions between a mflo/mfhi and the next mul/dmult/div/ddiv/divu/ddivu instruction. 
Fix issue2: In mips1-4, should not put mflo into the delay slot for the return.

Fix https://github.com/llvm/llvm-project/issues/81291

### CODE REVIEW

**Comment by yingopq** [comment]
@topperc @wzssyqa  Could you help review? Thanks!

**Comment by yingopq** [comment]
Ping @wzssyqa

**Comment by yingopq** [comment]
ping @wzssyqa @topperc, thanks!

**Comment by brad0** [comment]
@wzssyqa @topperc Ping.

**Comment by brad0** [comment]
@wzssyqa @topperc Ping.

**Comment by yingopq** [comment]
Hi @wzssyqa @topperc Could you help review? Thanks!

File: llvm/lib/Target/Mips/MipsBranchExpansion.cpp [hunk 1]
```diff
@@ -766,17 +769,40 @@ bool MipsBranchExpansion::handleSlot(Pred Predicate, Safe SafeInSlot) {
         if (std::next(Iit) == FI->end() ||
             std::next(Iit)->getOpcode() != Mips::NOP) {
           Changed = true;
-          TII->insertNop(*(I->getParent()), std::next(I), I->getDebugLoc())
-              ->bundleWithPred();
-          NumInsertedNops++;
+          if (Predicate(*I) ||
```
Comment (wzssyqa) [inline llvm/lib/Target/Mips/MipsBranchExpansion.cpp:772]: Why this condition is added here?
I guess you need some comment or description in commit msg.

File: llvm/lib/Target/Mips/MipsDelaySlotFiller.cpp [hunk 2]
```diff
@@ -743,6 +743,12 @@ bool MipsDelaySlotFiller::searchRange(MachineBasicBlock &MBB, IterTy Begin,
     bool InMicroMipsMode = STI.inMicroMipsMode();
     const MipsInstrInfo *TII = STI.getInstrInfo();
     unsigned Opcode = (*Slot).getOpcode();
+
+    if ((CurrI->getOpcode() == Mips::MFLO ||
```
Comment (wzssyqa) [inline llvm/lib/Target/Mips/MipsDelaySlotFiller.cpp:747]: No MFHI here?

File: llvm/lib/Target/Mips/MipsBranchExpansion.cpp (hunk 1 above)
Comment (yingopq) [reply llvm/lib/Target/Mips/MipsBranchExpansion.cpp:772]: I would check again without this condition and add comment.

Comment (yingopq) [reply llvm/lib/Target/Mips/MipsBranchExpansion.cpp:772]: I deleted this condition and the test result of ./build/bin/clang --target=mipsel-freestanding -mcpu=mips3 -fomit-frame-pointer -S 4.c -o 1.s` or without `-O3` all are OK, I would update it. Maybe some reasons I added it now I forget. Thanks!

File: llvm/lib/Target/Mips/MipsDelaySlotFiller.cpp (hunk 2 above)
Comment (yingopq) [reply llvm/lib/Target/Mips/MipsDelaySlotFiller.cpp]: Updated.

File: llvm/lib/Target/Mips/MipsBranchExpansion.cpp (hunk 1 above)
Comment (yingopq) [reply llvm/lib/Target/Mips/MipsBranchExpansion.cpp]: Updated. I added a new function which makes it simpler.

Comment (yingopq) [reply llvm/lib/Target/Mips/MipsBranchExpansion.cpp]: @wzssyqa Could you help review, thanks!

**Pull request merged**